sdk/python/kfp/dsl/component_spec.py

```diff
@@ -107,38 +110,118 @@ def build_task_inputs_spec(
     pipeline_params: The list of pipeline params.
     tasks_in_current_dag: The list of tasks names for tasks in the same dag.
+    is_parent_component_root: Whether the task is in the root component.
   """
   for param in pipeline_params or []:
-    input_name = param.full_name
+
+    input_name = _pipeline_param.additional_input_name_for_pipelineparam(param)
     if type_utils.is_parameter_type(param.param_type):
-      if param.op_name in tasks_in_current_dag:
+      if param.op_name and dsl_utils.sanitize_task_name(
+          param.op_name) in tasks_in_current_dag:
         task_spec.inputs.parameters[
             input_name].task_output_parameter.producer_task = (
                 dsl_utils.sanitize_task_name(param.op_name))
         task_spec.inputs.parameters[
             input_name].task_output_parameter.output_parameter_key = (
                 param.name)
       else:
-        task_spec.inputs.parameters[
-            input_name].component_input_parameter = input_name
+        task_spec.inputs.parameters[input_name].component_input_parameter = (
+            param.full_name if is_parent_component_root else input_name)
     else:
-      if param.op_name in tasks_in_current_dag:
+      if param.op_name and dsl_utils.sanitize_task_name(
+          param.op_name) in tasks_in_current_dag:
         task_spec.inputs.artifacts[
             input_name].task_output_artifact.producer_task = (
                 dsl_utils.sanitize_task_name(param.op_name))
         task_spec.inputs.artifacts[
             input_name].task_output_artifact.output_artifact_key = (
                 param.name)
       else:
-        task_spec.inputs.artifacts[
-            input_name].component_input_artifact = input_name
+        task_spec.inputs.artifacts[input_name].component_input_artifact = (
+            param.full_name if is_parent_component_root else input_name)
+
+
+def update_task_inputs_spec(
+    task_spec: pipeline_spec_pb2.PipelineTaskSpec,
+    parent_component_inputs: pipeline_spec_pb2.ComponentInputsSpec,
+    pipeline_params: List[_pipeline_param.PipelineParam],
+    tasks_in_current_dag: List[str],
+) -> None:
+  """Updates task inputs spec.
+
+  A task input may reference an output outside its immediate DAG.
+  For instance::
+
+    random_num = random_num_op(...)
+    with dsl.Condition(random_num.output > 5):
+      print_op('%s > 5' % random_num.output)
+
+  In this example, `dsl.Condition` forms a sub-DAG with one task from `print_op`
+  inside the sub-DAG. The task of `print_op` references output from `random_num`
+  task, which is outside the sub-DAG. When compiling to IR, such cross DAG
+  reference is disallowed. So we need to "punch a hole" in the sub-DAG to make
+  the input available in the sub-DAG component inputs if it's not already there,
+  Next, we can cal this method to fix the tasks inside the sub-DAG to make them
+  reference the component inputs instead of directly referening the original
+  producer task.
```

Comment:
This comment is useful at describing the problem. Can you also expand a bit on where this hole punching occurs and what it looks like?

Reply:
Btw, I get the sense that this is needed because the compiler "pattern" is taken after the v1 compiler. Can we mitigate this if we wrote this with a cleaner abstraction?

Reply:
Although the v1 compiler suffers from this issue, I'm not sure the v2 compiler "inherited" it.

The issue arises, because the DAGs are implicit in the DSL.
To get data in to or out of the inner DAG, we need to "punch holes" in all of the intermediate DAGs.
And these DAG inputs and outputs need to have unique names.

Reply:
BTW, the `dsl.Pipeline` compilation context and `get_current_pipeline` exist, so it's technically possible to know the DAG hierarchy. But I'm not sure it's a good idea to use them.